From: Andrew Morton <akpm@linux-foundation.org>
Subject: selinux-use-vma_is_stack-and-vma_is_heap-fix
Date: Wed Jul 12 01:42:33 PM PDT 2023

fix build

security/selinux/hooks.c: In function 'selinux_file_mprotect':
security/selinux/hooks.c:3768:42: error: suggest parentheses around '&&' within '||' [-Werror=parentheses]
 3768 |                 } else if (!vma->vm_file && vma_is_stack(vma) ||
      |                            ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
security/selinux/hooks.c:3769:59: error: expected statement before ')' token
 3769 |                             vma_is_stack_for_current(vma))) {
      |                                                           ^
security/selinux/hooks.c:3768:24: error: this 'if' clause does not guard... [-Werror=misleading-indentation]
 3768 |                 } else if (!vma->vm_file && vma_is_stack(vma) ||
      |                        ^~
security/selinux/hooks.c:3769:61: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the 'if'
 3769 |                             vma_is_stack_for_current(vma))) {
      |                                                             ^
security/selinux/hooks.c:3772:19: error: expected '}' before 'else'
 3772 |                 } else if (vma->vm_file && vma->anon_vma) {
      |                   ^~~~
security/selinux/hooks.c:3780:25: error: 'rc' undeclared (first use in this function); did you mean 'rq'?
 3780 |                         rc = file_has_perm(cred, vma->vm_file, FILE__EXECMOD);
      |                         ^~
      |                         rq
security/selinux/hooks.c:3780:25: note: each undeclared identifier is reported only once for each function it appears in
security/selinux/hooks.c: At top level:
security/selinux/hooks.c:3786:9: error: expected identifier or '(' before 'return'
 3786 |         return file_map_prot_check(vma->vm_file, prot, vma->vm_flags&VM_SHARED);
      |         ^~~~~~
security/selinux/hooks.c:3787:1: error: expected identifier or '(' before '}' token
 3787 | }
      | ^
security/selinux/hooks.c: In function 'selinux_file_mprotect':
security/selinux/hooks.c:3784:9: error: control reaches end of non-void function [-Werror=return-type]
 3784 |         }
      |         ^

Cc: Eric Paris <eparis@parisplace.org>
Cc: Kefeng Wang <wangkefeng.wang@huawei.com>
Cc: Paul Moore <paul@paul-moore.com>
Cc: Stephen Smalley <stephen.smalley.work@gmail.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 security/selinux/hooks.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

--- a/security/selinux/hooks.c~selinux-use-vma_is_stack-and-vma_is_heap-fix
+++ a/security/selinux/hooks.c
@@ -3765,8 +3765,9 @@ static int selinux_file_mprotect(struct
 		if (vma_is_heap(vma)) {
 			rc = avc_has_perm(sid, sid, SECCLASS_PROCESS,
 					  PROCESS__EXECHEAP, NULL);
-		} else if (!vma->vm_file && vma_is_stack(vma) ||
-			    vma_is_stack_for_current(vma))) {
+		} else if (!vma->vm_file &&
+				(vma_is_stack(vma) ||
+				 vma_is_stack_for_current(vma))) {
 			rc = avc_has_perm(sid, sid, SECCLASS_PROCESS,
 					  PROCESS__EXECSTACK, NULL);
 		} else if (vma->vm_file && vma->anon_vma) {
_
